pkg/discovery/oci_central.go

```diff
@@ -0,0 +1,250 @@
+	name               string
+	target             string
+	recommendedVersion string
+	version            string
+	hidden             string
+	description        string
+	publisher          string
+	vendor             string
+	os                 string
+	arch               string
+	digest             string
+	uri                string
+}
+
+// Name of the discovery.
+func (od *OCIDiscoveryForCentralRepo) Name() string {
+	return od.name
+}
+
+// Type of the discovery.
+func (od *OCIDiscoveryForCentralRepo) Type() string {
+	return common.DiscoveryTypeOCI
+}
+
+// List available plugins.
+func (od *OCIDiscoveryForCentralRepo) List() (plugins []Discovered, err error) {
+	dbFilePath, err := od.fetchContentDB()
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to fetch the content of discovery '%s'", od.Name())
+	}
+
+	return od.getPluginsFromDB(dbFilePath)
+}
+
+// Describe a plugin.
+func (od *OCIDiscoveryForCentralRepo) Describe(name string) (Discovered, error) {
+	return Discovered{}, errors.Errorf("NOT IMPLEMENTED")
+}
+
+// fetchContentDB downloads the content DB to the cache directory and returns
+// the path to the DB file to use.
+func (od *OCIDiscoveryForCentralRepo) fetchContentDB() (string, error) {
+	// TODO(khouzam): Improve by checking if we really need to download again or if we can use the cache
+	pluginDBDir := filepath.Join(common.DefaultCacheDir, "plugin_db")
+	if err := carvelhelpers.DownloadDBImage(od.image, pluginDBDir); err != nil {
+		return "", errors.Wrapf(err, "failed to download OCI image from discovery '%s'", od.Name())
+	}
+	return filepath.Join(pluginDBDir, common.CentralRepoDBFileName), nil
+}
+
+// getPluginsFromDB returns all plugins found in the DB located at the specified dbFilePath
+func (od *OCIDiscoveryForCentralRepo) getPluginsFromDB(dbFilePath string) ([]Discovered, error) {
+	db, err := sql.Open("sqlite3", dbFilePath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open the DB for discovery '%s'", od.Name())
+	}
+	defer db.Close()
+
+	// We need to order the results properly because the logic of processRows()
+	// expects an ordering of PluginName, then Target, then Version.
+	// The column order must also match the order used in getNextRow().
+	dbQuery := "SELECT PluginName,Target,RecommendedVersion,Version,Hidden,Description,Publisher,Vendor,OS,Architecture,Digest,URI FROM PluginBinaries ORDER BY PluginName,Target,Version;"
+	rows, err := db.Query(dbQuery)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to setup DB query for discovery '%s'", od.Name())
+	}
+	defer rows.Close()
+
+	allPluginPtrs, err := od.processRows(rows)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert from plugin pointers to plugins
+	// TODO(khouzam): continue optimizing by converting every call
+	// to using pointers
+	var allPlugins []Discovered
+	for _, pluginPtr := range allPluginPtrs {
+		allPlugins = append(allPlugins, *pluginPtr)
+	}
+	return allPlugins, nil
+}
```

Comment:
Should we put this into a separate golang package that handles all the SQL queries together?
We need to add more SQL queries that can be used at multiple places. Combining all SQL queries related stuff into separate packages might be better. WDYT? @marckhouzam

Reply:
That is a good idea.  I will work something out and update.

Reply:
+1 to factoring out the db access from the logic of obtaining the 'db' via the OCI registry to minimize churn should we elect to use a different db backend in the future.

Reply:
I've pushed a commit to address this.

The approach I took is to create a new `DiscoveryBackend` interface. A `DiscoveryBackend` represents the content management logic behind a plugin `Discovery`.  The commit also creates an implementation of this new `DiscoveryBackend` for the SQLite backend of the Central Repository.

Note: The `DiscoveryBackend` support is included in the "discovery" package itself at the moment.  This is so that it can use the `Discovered` type.  Making the new `DiscoveryBackend` part of its own package would prevent it from using the `Discovered` type due to a cyclical dependency.  This would then force us to create yet another type to characterize plugins for the `DiscoveryBackend` to use.  This new type would then need to be converted to the `Discovered` type by the `Discovery`.

An alternative would be to move the `Discovered` type to its own package that could be imported by both the "discovery" package, and a new "discoverybackend" package.

We should discuss this.

Reply:
We have discussed this and we have agreed that we will create a new type that the `DiscoveryBackend` can return, so as to decouple the `DiscoveryBackend` from the `discovery` package.
This will be done in a follow-up PR to allow this PR to get in now.